lib/Reader/readerir.cpp

```diff
@@ -3082,48 +3083,53 @@ Type *GenIR::binaryOpType(ReaderBaseNS::BinaryOpcode Opcode, Type *Type1,
-  const bool Type2IsUnmanagedPointer = isUnmanagedPointerType(Type2);
-  const bool IsStrictlyAdd = (Opcode == ReaderBaseNS::Add);
-  const bool IsAdd = IsStrictlyAdd || (Opcode == ReaderBaseNS::AddOvf) ||
-                     (Opcode == ReaderBaseNS::AddOvfUn);
-  const bool IsStrictlySub = (Opcode == ReaderBaseNS::Sub);
-  const bool IsSub = IsStrictlySub || (Opcode == ReaderBaseNS::SubOvf) ||
-                     (Opcode == ReaderBaseNS::SubOvfUn);
-  const bool IsStrictlyAddOrSub = IsStrictlyAdd || IsStrictlySub;
-  const bool IsAddOrSub = IsAdd || IsSub;
-
-  if (Type1IsUnmanagedPointer || Type2IsUnmanagedPointer) {
-    // ptr +/- int = ptr
-    if (IsAddOrSub && Type1IsUnmanagedPointer && Type2IsInt &&
-        (Size1 >= Size2)) {
-      return Type1;
-    }
-    // int + ptr = ptr
-    if (IsAdd && Type1IsInt && Type2IsUnmanagedPointer && (Size2 >= Size1)) {
-      return Type2;
-    }
-    // Otherwise type result as native int as long as there's no truncation
-    // going on.
-    if ((Size1 <= TargetPointerSizeInBits) &&
-        (Size2 <= TargetPointerSizeInBits)) {
-      return Type::getIntNTy(*JitContext->LLVMContext, TargetPointerSizeInBits);
+  } else {
+    const bool Type1IsUnmanagedPointer = isUnmanagedPointerType(Type1);
+    const bool Type2IsUnmanagedPointer = isUnmanagedPointerType(Type2);
+    const bool IsStrictlyAdd = (Opcode == ReaderBaseNS::Add);
+    const bool IsAdd = IsStrictlyAdd || (Opcode == ReaderBaseNS::AddOvf) ||
+                       (Opcode == ReaderBaseNS::AddOvfUn);
+    const bool IsStrictlySub = (Opcode == ReaderBaseNS::Sub);
+    const bool IsSub = IsStrictlySub || (Opcode == ReaderBaseNS::SubOvf) ||
+                       (Opcode == ReaderBaseNS::SubOvfUn);
+    const bool IsStrictlyAddOrSub = IsStrictlyAdd || IsStrictlySub;
+    const bool IsAddOrSub = IsAdd || IsSub;
+
+    // If we see a mixture of int and unmanaged pointer, the result
+    // is generally a native int, with a few special cases where we
+    // preserve pointer-ness.
+    if (Type1IsUnmanagedPointer || Type2IsUnmanagedPointer) {
+      // ptr +/- int = ptr
+      if (IsAddOrSub && Type1IsUnmanagedPointer && Type2IsInt &&
+          (Size1 >= Size2)) {
+        return Type1;
+      }
+      // int + ptr = ptr
+      if (IsAdd && Type1IsInt && Type2IsUnmanagedPointer && (Size2 >= Size1)) {
+        return Type2;
+      }
+      // Otherwise type result as native int as long as there's no truncation
+      // going on.
+      if ((Size1 <= TargetPointerSizeInBits) &&
+          (Size2 <= TargetPointerSizeInBits)) {
+        return Type::getIntNTy(*JitContext->LLVMContext,
+                               TargetPointerSizeInBits);
+      }
+    } else if (isManagedPointerType(Type1)) {
+      if (IsSub && isManagedPointerType(Type2)) {
+        // The difference of two managed pointers is a native int.
+        return Type::getIntNTy(*JitContext->LLVMContext,
+                               TargetPointerSizeInBits);
```

Comment:
The caller is going to create `PtrToInt`s over the two sources and then do an integer subtract.  Moving one `PtrToInt` across a GC safepoint but not the other would be bad.  Does safepoint insertion run before any optimizations might try to do that?